File: tradewizard-agents/src/nodes/cross-examination.ts
/**
 * Cross-Examination Node
 *
 * This module implements the cross-examination stage of the debate protocol.
 * It challenges thesis assumptions through structured tests using LLM analysis.
 */

import { createLLMInstance, type LLMInstance } from '../utils/llm-factory.js';
import type { GraphStateType } from '../models/state.js';
import type { DebateRecord, DebateTest, Thesis } from '../models/types.js';
import type { EngineConfig } from '../config/index.js';
import { formatTimestamp } from '../utils/timestamp-formatter.js';

/**
 * Type for supported LLM instances
 */
// Removed - now imported from llm-factory

/**
 * Create LLM instance for cross-examination
 *
 * In single-provider mode: use the configured LLM
 * In multi-provider mode: use default LLM (ChatOpenAI with GPT-4-turbo)
 *
 * @param config - Engine configuration
 * @returns LLM instance for cross-examination
 */
function createCrossExaminationLLM(config: EngineConfig): LLMInstance {
  // Use configured LLM respecting single/multi provider mode
  // In multi-provider mode, prefer OpenAI for cross-examination (good at adversarial reasoning)
  return createLLMInstance(config, 'openai', ['anthropic', 'google', 'nova']);
}

/**
 * System prompt for evidence testing
 */
const EVIDENCE_TEST_PROMPT = `Current date and time: ${formatTimestamp(Date.now()).formatted}

You are a fact-checker for prediction market theses.

Your role is to verify factual claims made in a thesis against available data.

Given a thesis and its claims, evaluate:
- Are the factual claims verifiable?
- Is the evidence cited accurate and relevant?
- Are there contradictory facts that weaken the claim?

For each claim, determine if it:
- SURVIVED: The claim is well-supported by evidence
- WEAKENED: The claim has some support but also contradictory evidence
- REFUTED: The claim is contradicted by available evidence

Respond with a structured test result including:
- testType: "evidence"
- claim: The specific claim being tested
- challenge: Your analysis of the evidence
- outcome: "survived", "weakened", or "refuted"
- score: A number from -1 (refuted) to 1 (survived), with 0 being neutral`;

/**
 * System prompt for causality testing
 */
const CAUSALITY_TEST_PROMPT = `Current date and time: ${formatTimestamp(Date.now()).formatted}

You are a causality analyst for prediction market theses.

Your role is to test whether claimed correlations actually imply causation.

Given a thesis and its causal claims, evaluate:
- Does the thesis confuse correlation with causation?
- Are there alternative explanations for the observed correlation?
- Is the causal mechanism plausible and well-explained?

For each causal claim, determine if it:
- SURVIVED: The causal claim is well-reasoned with a clear mechanism
- WEAKENED: The causal claim is plausible but has alternative explanations
- REFUTED: The claim confuses correlation with causation

Respond with a structured test result including:
- testType: "causality"
- claim: The specific causal claim being tested
- challenge: Your analysis of the causality
- outcome: "survived", "weakened", or "refuted"
- score: A number from -1 (refuted) to 1 (survived), with 0 being neutral`;

/**
 * System prompt for timing testing
 */
const TIMING_TEST_PROMPT = `Current date and time: ${formatTimestamp(Date.now()).formatted}

You are a timeline analyst for prediction market theses.

Your role is to validate catalyst timelines and event sequences.

Given a thesis and its catalysts, evaluate:
- Are the catalyst timelines realistic?
- Is there enough time for the catalysts to have their claimed effect?
- Are the event sequences logically ordered?

For each timing claim, determine if it:
- SURVIVED: The timeline is realistic and well-reasoned
- WEAKENED: The timeline is plausible but tight or uncertain
- REFUTED: The timeline is unrealistic or contradictory

Respond with a structured test result including:
- testType: "timing"
- claim: The specific timing claim being tested
- challenge: Your analysis of the timeline
- outcome: "survived", "weakened", or "refuted"
- score: A number from -1 (refuted) to 1 (survived), with 0 being neutral`;

/**
 * System prompt for liquidity testing
 */
const LIQUIDITY_TEST_PROMPT = `Current date and time: ${formatTimestamp(Date.now()).formatted}

You are a market microstructure analyst for prediction market theses.

Your role is to assess execution feasibility given market liquidity.

Given a thesis and market conditions, evaluate:
- Can the recommended position be executed at stated prices?
- Is there sufficient liquidity to enter and exit without significant slippage?
- Are there liquidity risks that could impact the trade?

For the liquidity assessment, determine if it:
- SURVIVED: Liquidity is sufficient for execution
- WEAKENED: Liquidity is adequate but with some slippage risk
- REFUTED: Liquidity is insufficient for the recommended position size

Respond with a structured test result including:
- testType: "liquidity"
- claim: The liquidity assumption being tested
- challenge: Your analysis of execution feasibility
- outcome: "survived", "weakened", or "refuted"
- score: A number from -1 (refuted) to 1 (survived), with 0 being neutral`;

/**
 * System prompt for tail risk testing
 */
const TAIL_RISK_TEST_PROMPT = `Current date and time: ${formatTimestamp(Date.now()).formatted}

You are a tail risk analyst for prediction market theses.

Your role is to identify low-probability, high-impact scenarios.

Given a thesis, evaluate:
- Are there low-probability scenarios that could invalidate the thesis?
- Has the thesis adequately considered tail risks?
- Are there "black swan" events that could dramatically change the outcome?

For the tail risk assessment, determine if it:
- SURVIVED: Tail risks are acknowledged and accounted for
- WEAKENED: Some tail risks are present but not fully addressed
- REFUTED: Critical tail risks are ignored or dismissed

Respond with a structured test result including:
- testType: "tail-risk"
- claim: The tail risk assumption being tested
- challenge: Your analysis of tail risk scenarios
- outcome: "survived", "weakened", or "refuted"
- score: A number from -1 (refuted) to 1 (survived), with 0 being neutral`;

/**
 * Execute evidence test on a thesis
 */
async function executeEvidenceTest(
  llm: LLMInstance,
  thesis: Thesis,
  opposingThesis: Thesis
): Promise<DebateTest> {
  const prompt = `Evaluate the factual claims in this thesis:

Thesis Direction: ${thesis.direction}
Core Argument: ${thesis.coreArgument}
Catalysts: ${thesis.catalysts.join(', ')}

Opposing Thesis: ${opposingThesis.coreArgument}

Analyze the factual claims and determine if they are well-supported by evidence.`;

  const response = await llm.invoke([
    { role: 'system', content: EVIDENCE_TEST_PROMPT },
    { role: 'user', content: prompt },
  ]);

  // Parse the response to extract test result
  const content = typeof response.content === 'string' ? response.content : JSON.stringify(response.content);
  
  // Simple parsing logic - look for keywords in the response
  let outcome: 'survived' | 'weakened' | 'refuted' = 'survived';
  const lowerContent = content.toLowerCase();
  
  if (lowerContent.includes('refuted') || lowerContent.includes('contradicted') || lowerContent.includes('false')) {
    outcome = 'refuted';
  } else if (lowerContent.includes('weakened') || lowerContent.includes('questionable') || lowerContent.includes('uncertain')) {
    outcome = 'weakened';
  }
  
  const score = outcome === 'survived' ? 0.7 : outcome === 'weakened' ? 0 : -0.7;

  return {
    testType: 'evidence',
    claim: thesis.coreArgument,
    challenge: content.substring(0, 200), // Truncate for brevity
    outcome,
    score,
  };
}

/**
 * Execute causality test on a thesis
 */
async function executeCausalityTest(
  llm: LLMInstance,
  thesis: Thesis,
  opposingThesis: Thesis
): Promise<DebateTest> {
  const prompt = `Evaluate the causal claims in this thesis:

Thesis Direction: ${thesis.direction}
Core Argument: ${thesis.coreArgument}
Catalysts: ${thesis.catalysts.join(', ')}

Opposing Thesis: ${opposingThesis.coreArgument}

Analyze whether the thesis confuses correlation with causation.`;

  const response = await llm.invoke([
    { role: 'system', content: CAUSALITY_TEST_PROMPT },
    { role: 'user', content: prompt },
  ]);

  const content = typeof response.content === 'string' ? response.content : JSON.stringify(response.content);
  
  let outcome: 'survived' | 'weakened' | 'refuted' = 'survived';
  const lowerContent = content.toLowerCase();
  
  if (lowerContent.includes('refuted') || lowerContent.includes('correlation') && lowerContent.includes('not') && lowerContent.includes('causation')) {
    outcome = 'refuted';
  } else if (lowerContent.includes('weakened') || lowerContent.includes('alternative')) {
    outcome = 'weakened';
  }
  
  const score = outcome === 'survived' ? 0.7 : outcome === 'weakened' ? 0 : -0.7;

  return {
    testType: 'causality',
    claim: thesis.coreArgument,
    challenge: content.substring(0, 200),
    outcome,
    score,
  };
}

/**
 * Execute timing test on a thesis
 */
async function executeTimingTest(
  llm: LLMInstance,
  thesis: Thesis,
  mbd: GraphStateType['mbd']
): Promise<DebateTest> {
  if (!mbd) {
    return {
      testType: 'timing',
      claim: 'Catalyst timeline',
      challenge: 'No market data available for timing analysis',
      outcome: 'weakened',
      score: 0,
    };
  }

  const marketExpiry = formatTimestamp(mbd.expiryTimestamp);
  const prompt = `Evaluate the timing of catalysts in this thesis:

Thesis Direction: ${thesis.direction}
Catalysts: ${thesis.catalysts.join(', ')}
Market Expiry: ${marketExpiry.formatted}

Analyze whether the catalyst timeline is realistic given the market expiry.`;

  const response = await llm.invoke([
    { role: 'system', content: TIMING_TEST_PROMPT },
    { role: 'user', content: prompt },
  ]);

  const content = typeof response.content === 'string' ? response.content : JSON.stringify(response.content);
  
  let outcome: 'survived' | 'weakened' | 'refuted' = 'survived';
  const lowerContent = content.toLowerCase();
  
  if (lowerContent.includes('refuted') || lowerContent.includes('unrealistic') || lowerContent.includes('impossible')) {
    outcome = 'refuted';
  } else if (lowerContent.includes('weakened') || lowerContent.includes('tight') || lowerContent.includes('uncertain')) {
    outcome = 'weakened';
  }
  
  const score = outcome === 'survived' ? 0.7 : outcome === 'weakened' ? 0 : -0.7;

  return {
    testType: 'timing',
    claim: `Catalysts: ${thesis.catalysts.join(', ')}`,
    challenge: content.substring(0, 200),
    outcome,
    score,
  };
}

/**
 * Execute liquidity test on a thesis
 */
async function executeLiquidityTest(
  llm: LLMInstance,
  thesis: Thesis,
  mbd: GraphStateType['mbd']
): Promise<DebateTest> {
  if (!mbd) {
    return {
      testType: 'liquidity',
      claim: 'Execution feasibility',
      challenge: 'No market data available for liquidity analysis',
      outcome: 'weakened',
      score: 0,
    };
  }

  const prompt = `Evaluate the execution feasibility of this thesis:

Thesis Direction: ${thesis.direction}
Market Liquidity Score: ${mbd.liquidityScore}/10
Bid-Ask Spread: ${mbd.bidAskSpread} cents
24h Volume: $${mbd.volume24h}

Analyze whether the position can be executed without significant slippage.`;

  const response = await llm.invoke([
    { role: 'system', content: LIQUIDITY_TEST_PROMPT },
    { role: 'user', content: prompt },
  ]);

  const content = typeof response.content === 'string' ? response.content : JSON.stringify(response.content);
  
  let outcome: 'survived' | 'weakened' | 'refuted' = 'survived';
  const lowerContent = content.toLowerCase();
  
  if (lowerContent.includes('refuted') || lowerContent.includes('insufficient') || lowerContent.includes('cannot')) {
    outcome = 'refuted';
  } else if (lowerContent.includes('weakened') || lowerContent.includes('slippage') || lowerContent.includes('risk')) {
    outcome = 'weakened';
  }
  
  const score = outcome === 'survived' ? 0.7 : outcome === 'weakened' ? 0 : -0.7;

  return {
    testType: 'liquidity',
    claim: 'Sufficient liquidity for execution',
    challenge: content.substring(0, 200),
    outcome,
    score,
  };
}

/**
 * Execute tail risk test on a thesis
 */
async function executeTailRiskTest(
  llm: LLMInstance,
  thesis: Thesis
): Promise<DebateTest> {
  const prompt = `Evaluate the tail risk considerations in this thesis:

Thesis Direction: ${thesis.direction}
Core Argument: ${thesis.coreArgument}
Failure Conditions: ${thesis.failureConditions.join(', ')}

Analyze whether the thesis adequately considers low-probability, high-impact scenarios.`;

  const response = await llm.invoke([
    { role: 'system', content: TAIL_RISK_TEST_PROMPT },
    { role: 'user', content: prompt },
  ]);

  const content = typeof response.content === 'string' ? response.content : JSON.stringify(response.content);
  
  let outcome: 'survived' | 'weakened' | 'refuted' = 'survived';
  const lowerContent = content.toLowerCase();
  
  if (lowerContent.includes('refuted') || lowerContent.includes('ignored') || lowerContent.includes('dismissed')) {
    outcome = 'refuted';
  } else if (lowerContent.includes('weakened') || lowerContent.includes('not fully') || lowerContent.includes('inadequate')) {
    outcome = 'weakened';
  }
  
  const score = outcome === 'survived' ? 0.7 : outcome === 'weakened' ? 0 : -0.7;

  return {
    testType: 'tail-risk',
    claim: `Failure conditions: ${thesis.failureConditions.join(', ')}`,
    challenge: content.substring(0, 200),
    outcome,
    score,
  };
}

/**
 * Calculate aggregate scores from debate tests
 */
function calculateDebateScores(tests: DebateTest[]): {
  bullScore: number;
  bearScore: number;
} {
  // Tests are alternated: bull tests, then bear tests
  // For simplicity, we'll calculate average scores
  const bullTests = tests.filter((_, index) => index % 2 === 0);
  const bearTests = tests.filter((_, index) => index % 2 === 1);

  const bullScore =
    bullTests.length > 0
      ? bullTests.reduce((sum, test) => sum + test.score, 0) / bullTests.length
      : 0;

  const bearScore =
    bearTests.length > 0
      ? bearTests.reduce((sum, test) => sum + test.score, 0) / bearTests.length
      : 0;

  return { bullScore, bearScore };
}

/**
 * Identify key disagreements between theses
 */
function identifyKeyDisagreements(
  bullThesis: Thesis,
  bearThesis: Thesis,
  tests: DebateTest[]
): string[] {
  const disagreements: string[] = [];

  // Check probability disagreement
  const probDiff = Math.abs(bullThesis.fairProbability - bearThesis.fairProbability);
  if (probDiff > 0.2) {
    disagreements.push(
      `Significant probability disagreement: Bull ${(bullThesis.fairProbability * 100).toFixed(1)}% vs Bear ${(bearThesis.fairProbability * 100).toFixed(1)}%`
    );
  }

  // Check for refuted tests
  const refutedTests = tests.filter((test) => test.outcome === 'refuted');
  if (refutedTests.length > 0) {
    disagreements.push(
      `${refutedTests.length} claims were refuted during cross-examination`
    );
  }

  // Check for conflicting catalysts
  const bullCatalysts = new Set(bullThesis.catalysts);
  const bearCatalysts = new Set(bearThesis.catalysts);
  const hasConflictingCatalysts =
    bullCatalysts.size > 0 &&
    bearCatalysts.size > 0 &&
    ![...bullCatalysts].some((c) => bearCatalysts.has(c));

  if (hasConflictingCatalysts) {
    disagreements.push('Theses identify different key catalysts');
  }

  return disagreements;
}

/**
 * Create cross-examination node factory
 *
 * This factory function creates a cross-examination node with the configured LLM.
 *
 * @param config - Engine configuration
 * @returns Cross-examination node function
 */
export function createCrossExaminationNode(
  config: EngineConfig
): (state: GraphStateType) => Promise<Partial<GraphStateType>> {
  const llm = createCrossExaminationLLM(config);

  return async (state: GraphStateType): Promise<Partial<GraphStateType>> => {
    const startTime = Date.now();

    // Check if theses are available
    if (!state.bullThesis || !state.bearThesis) {
      return {
        consensusError: {
          type: 'INSUFFICIENT_DATA',
          reason: 'Bull and bear theses are required for cross-examination',
        },
        auditLog: [
          {
            stage: 'cross_examination',
            timestamp: Date.now(),
            data: {
              success: false,
              error: 'Missing theses',
              hasBullThesis: !!state.bullThesis,
              hasBearThesis: !!state.bearThesis,
              duration: Date.now() - startTime,
            },
          },
        ],
      };
    }

    try {
      const tests: DebateTest[] = [];

      // Execute evidence test on bull thesis
      const bullEvidenceTest = await executeEvidenceTest(
        llm,
        state.bullThesis,
        state.bearThesis
      );
      tests.push(bullEvidenceTest);

      // Execute evidence test on bear thesis
      const bearEvidenceTest = await executeEvidenceTest(
        llm,
        state.bearThesis,
        state.bullThesis
      );
      tests.push(bearEvidenceTest);

      // Execute causality test on bull thesis
      const bullCausalityTest = await executeCausalityTest(
        llm,
        state.bullThesis,
        state.bearThesis
      );
      tests.push(bullCausalityTest);

      // Execute causality test on bear thesis
      const bearCausalityTest = await executeCausalityTest(
        llm,
        state.bearThesis,
        state.bullThesis
      );
      tests.push(bearCausalityTest);

      // Execute timing test on bull thesis
      const bullTimingTest = await executeTimingTest(llm, state.bullThesis, state.mbd);
      tests.push(bullTimingTest);

      // Execute timing test on bear thesis
      const bearTimingTest = await executeTimingTest(llm, state.bearThesis, state.mbd);
      tests.push(bearTimingTest);

      // Execute liquidity test on bull thesis
      const bullLiquidityTest = await executeLiquidityTest(
        llm,
        state.bullThesis,
        state.mbd
      );
      tests.push(bullLiquidityTest);

      // Execute liquidity test on bear thesis
      const bearLiquidityTest = await executeLiquidityTest(
        llm,
        state.bearThesis,
        state.mbd
      );
      tests.push(bearLiquidityTest);

      // Execute tail risk test on bull thesis
      const bullTailRiskTest = await executeTailRiskTest(llm, state.bullThesis);
      tests.push(bullTailRiskTest);

      // Execute tail risk test on bear thesis
      const bearTailRiskTest = await executeTailRiskTest(llm, state.bearThesis);
      tests.push(bearTailRiskTest);

      // Calculate aggregate scores
      const { bullScore, bearScore } = calculateDebateScores(tests);

      // Identify key disagreements
      const keyDisagreements = identifyKeyDisagreements(
        state.bullThesis,
        state.bearThesis,
        tests
      );

      // Create debate record
      const debateRecord: DebateRecord = {
        tests,
        bullScore,
        bearScore,
        keyDisagreements,
      };

      return {
        debateRecord,
        auditLog: [
          {
            stage: 'cross_examination',
            timestamp: Date.now(),
            data: {
              success: true,
              testsExecuted: tests.length,
              bullScore,
              bearScore,
              keyDisagreements: keyDisagreements.length,
              duration: Date.now() - startTime,
            },
          },
        ],
      };
    } catch (error) {
      return {
        consensusError: {
          type: 'CONSENSUS_FAILED',
          reason:
            error instanceof Error
              ? error.message
              : 'Unknown error during cross-examination',
        },
        auditLog: [
          {
            stage: 'cross_examination',
            timestamp: Date.now(),
            data: {
              success: false,
              error: error instanceof Error ? error.message : 'Unknown error',
              duration: Date.now() - startTime,
            },
          },
        ],
      };
    }
  };
}

/**
 * Default cross-examination node
 *
 * This is a convenience export that uses the default configuration.
 * For production use, create a node with createCrossExaminationNode(config).
 */
export const crossExaminationNode = (config: EngineConfig) =>
  createCrossExaminationNode(config);

File: doa/nodes/cross_examination.py
"""Cross-examination node for LangGraph workflow."""

import logging
import time
from typing import Any, Dict, List, Literal

from models.state import GraphState
from models.types import (
    AuditEntry,
    DebateRecord,
    DebateTest,
    MarketBriefingDocument,
    Thesis,
)
from config import EngineConfig
from utils.llm_factory import create_llm_instance
from prompts import CROSS_EXAMINATION_PROMPT

logger = logging.getLogger(__name__)


# Type for test outcomes
TestOutcome = Literal["survived", "weakened", "refuted"]


def outcome_to_score(outcome: TestOutcome) -> float:
    """
    Convert test outcome to numerical score.
    
    Args:
        outcome: Test outcome (survived/weakened/refuted)
        
    Returns:
        Score: +1.0 for survived, 0.0 for weakened, -1.0 for refuted
        
    Examples:
        >>> outcome_to_score("survived")
        1.0
        >>> outcome_to_score("weakened")
        0.0
        >>> outcome_to_score("refuted")
        -1.0
    """
    score_map = {
        "survived": 1.0,
        "weakened": 0.0,
        "refuted": -1.0
    }
    return score_map.get(outcome, 0.0)


def calculate_thesis_score(tests: List[DebateTest]) -> float:
    """
    Calculate overall thesis score from test results.
    
    Score is the average of all test scores.
    
    Args:
        tests: List of debate tests for a thesis
        
    Returns:
        Overall score (-1.0 to +1.0)
        
    Examples:
        >>> tests = [
        ...     DebateTest(outcome="survived", score=1.0, ...),
        ...     DebateTest(outcome="weakened", score=0.0, ...),
        ...     DebateTest(outcome="survived", score=1.0, ...)
        ... ]
        >>> score = calculate_thesis_score(tests)
        >>> assert score == (1.0 + 0.0 + 1.0) / 3
    """
    if not tests:
        return 0.0
    
    total_score = sum(test.score for test in tests)
    return total_score / len(tests)


def create_test_for_thesis(
    thesis: Thesis,
    test_type: Literal["evidence", "causality", "timing", "liquidity", "tail-risk"],
    mbd: MarketBriefingDocument
) -> Dict[str, Any]:
    """
    Create a test context for a specific thesis and test type.
    
    Args:
        thesis: The thesis to test
        test_type: Type of test to perform
        mbd: Market Briefing Document for context
        
    Returns:
        Dictionary with test context
        
    Examples:
        >>> thesis = Thesis(
        ...     direction="YES",
        ...     core_argument="Strong evidence supports YES",
        ...     catalysts=["Catalyst A"],
        ...     ...
        ... )
        >>> test = create_test_for_thesis(thesis, "evidence", mbd)
        >>> assert test["test_type"] == "evidence"
    """
    return {
        "test_type": test_type,
        "thesis_direction": thesis.direction,
        "core_argument": thesis.core_argument,
        "catalysts": thesis.catalysts,
        "failure_conditions": thesis.failure_conditions,
        "fair_probability": thesis.fair_probability,
        "market_probability": thesis.market_probability,
        "edge": thesis.edge,
        "supporting_signals": thesis.supporting_signals,
        "market_question": mbd.question,
        "event_type": mbd.event_type,
        "time_to_resolution": mbd.expiry_timestamp - int(time.time()),
        "liquidity_score": mbd.liquidity_score,
        "volatility_regime": mbd.volatility_regime
    }


def identify_key_disagreements(
    bull_thesis: Thesis,
    bear_thesis: Thesis,
    bull_tests: List[DebateTest],
    bear_tests: List[DebateTest]
) -> List[str]:
    """
    Identify key disagreements between bull and bear theses.
    
    Disagreements are identified from:
    1. Contradictory claims in core arguments
    2. Different catalyst interpretations
    3. Conflicting failure conditions
    4. Test outcomes that diverge significantly
    
    Args:
        bull_thesis: Bull thesis
        bear_thesis: Bear thesis
        bull_tests: Test results for bull thesis
        bear_tests: Test results for bear thesis
        
    Returns:
        List of key disagreement descriptions
        
    Examples:
        >>> bull = Thesis(direction="YES", core_argument="X will happen", ...)
        >>> bear = Thesis(direction="NO", core_argument="X will not happen", ...)
        >>> disagreements = identify_key_disagreements(bull, bear, [], [])
        >>> assert len(disagreements) > 0
    """
    disagreements = []
    
    # Fundamental directional disagreement
    prob_diff = abs(bull_thesis.fair_probability - bear_thesis.fair_probability)
    if prob_diff > 0.2:
        disagreements.append(
            f"Fundamental probability disagreement: Bull thesis estimates "
            f"{bull_thesis.fair_probability:.1%} while bear thesis estimates "
            f"{bear_thesis.fair_probability:.1%} (difference: {prob_diff:.1%})"
        )
    
    # Catalyst disagreements
    bull_catalysts_set = set(bull_thesis.catalysts)
    bear_catalysts_set = set(bear_thesis.catalysts)
    
    # Check if bear catalysts contradict bull catalysts
    if bull_catalysts_set and bear_catalysts_set:
        disagreements.append(
            f"Catalyst interpretation differs: Bull thesis emphasizes "
            f"{bull_thesis.catalysts[0]}, while bear thesis focuses on "
            f"{bear_thesis.catalysts[0]}"
        )
    
    # Test outcome disagreements
    for bull_test, bear_test in zip(bull_tests, bear_tests):
        if bull_test.test_type == bear_test.test_type:
            score_diff = abs(bull_test.score - bear_test.score)
            if score_diff >= 1.5:  # Significant divergence
                disagreements.append(
                    f"{bull_test.test_type.capitalize()} test shows divergence: "
                    f"Bull thesis {bull_test.outcome}, bear thesis {bear_test.outcome}"
                )
    
    # Edge disagreements
    if bull_thesis.edge > 0.1 and bear_thesis.edge > 0.1:
        disagreements.append(
            f"Both theses claim significant edge: Bull edge {bull_thesis.edge:.1%}, "
            f"bear edge {bear_thesis.edge:.1%}, suggesting market mispricing from both perspectives"
        )
    
    # Limit to top 5 most important disagreements
    return disagreements[:5]


async def cross_examination_node(
    state: GraphState,
    config: EngineConfig
) -> Dict[str, Any]:
    """
    Execute cross-examination tests on bull and bear theses.
    
    This node performs adversarial testing of both theses through five
    structured tests:
    1. Evidence test - quality and verification of supporting evidence
    2. Causality test - soundness of causal logic and reasoning
    3. Timing test - realism of timing and sequencing assumptions
    4. Liquidity test - feasibility of execution in practice
    5. Tail risk test - robustness to extreme scenarios
    
    Each thesis is scored based on how well it withstands these challenges.
    
    Args:
        state: Current workflow state with theses and MBD
        config: Engine configuration
        
    Returns:
        State update with debate_record and audit entry
        
    State Requirements:
        - bull_thesis: Thesis for YES outcome (required)
        - bear_thesis: Thesis for NO outcome (required)
        - mbd: MarketBriefingDocument for context (required)
        
    State Updates:
        - debate_record: DebateRecord with test results and scores
        - audit_log: Audit entry for cross-examination stage
        
    Examples:
        >>> state = {
        ...     "bull_thesis": Thesis(direction="YES", ...),
        ...     "bear_thesis": Thesis(direction="NO", ...),
        ...     "mbd": MarketBriefingDocument(...)
        ... }
        >>> result = await cross_examination_node(state, config)
        >>> assert "debate_record" in result
    """
    start_time = time.time()
    
    # Extract required data from state
    bull_thesis = state.get("bull_thesis")
    bear_thesis = state.get("bear_thesis")
    mbd = state.get("mbd")
    
    # Validate inputs
    if not bull_thesis or not bear_thesis:
        logger.warning("Cross-examination called without both theses")
        return {
            "audit_log": [AuditEntry(
                stage="cross_examination",
                timestamp=int(time.time()),
                status="failed",
                details={
                    "error": "Both bull and bear theses required",
                    "has_bull": bull_thesis is not None,
                    "has_bear": bear_thesis is not None
                }
            )]
        }
    
    if not mbd:
        logger.error("Cross-examination called without MBD")
        return {
            "audit_log": [AuditEntry(
                stage="cross_examination",
                timestamp=int(time.time()),
                status="failed",
                details={
                    "error": "Market Briefing Document not available"
                }
            )]
        }
    
    logger.info("Executing cross-examination of bull and bear theses")
    
    try:
        # Define structured output schema for debate record
        from pydantic import BaseModel, Field
        from typing import List as TypingList
        
        class DebateTestOutput(BaseModel):
            test_type: Literal["evidence", "causality", "timing", "liquidity", "tail-risk"]
            claim: str
            challenge: str
            outcome: Literal["survived", "weakened", "refuted"]
            score: float = Field(ge=-1.0, le=1.0)
        
        class CrossExaminationOutput(BaseModel):
            bull_tests: TypingList[DebateTestOutput] = Field(min_length=5, max_length=5)
            bear_tests: TypingList[DebateTestOutput] = Field(min_length=5, max_length=5)
            bull_score: float = Field(ge=-1.0, le=1.0)
            bear_score: float = Field(ge=-1.0, le=1.0)
            key_disagreements: TypingList[str] = Field(min_length=1, max_length=5)
        
        # Create LLM instance with rotation support and structured output for cross-examination
        from utils.llm_rotation_manager import LLMRotationManager
        
        rotation_manager = None
        if len(config.llm.model_names) > 1:
            model_names_str = ",".join(config.llm.model_names)
            rotation_manager = LLMRotationManager(model_names_str)
            logger.info(f"[cross_examination] Created rotation manager with {len(config.llm.model_names)} models")
        
        llm = create_llm_instance(
            config.llm, 
            rotation_manager=rotation_manager,
            structured_output_model=CrossExaminationOutput
        )
        
        # Define test types
        test_types: List[Literal["evidence", "causality", "timing", "liquidity", "tail-risk"]] = [
            "evidence",
            "causality",
            "timing",
            "liquidity",
            "tail-risk"
        ]
        
        # Prepare context for LLM
        bull_context = create_test_for_thesis(bull_thesis, "evidence", mbd)
        bear_context = create_test_for_thesis(bear_thesis, "evidence", mbd)
        
        # Format prompt with thesis information
        prompt_context = f"""
BULL THESIS (YES):
Direction: {bull_thesis.direction}
Fair Probability: {bull_thesis.fair_probability:.1%}
Market Probability: {bull_thesis.market_probability:.1%}
Edge: {bull_thesis.edge:.1%}
Core Argument: {bull_thesis.core_argument}
Catalysts: {', '.join(bull_thesis.catalysts)}
Failure Conditions: {', '.join(bull_thesis.failure_conditions)}
Supporting Signals: {', '.join(bull_thesis.supporting_signals)}

BEAR THESIS (NO):
Direction: {bear_thesis.direction}
Fair Probability: {bear_thesis.fair_probability:.1%}
Market Probability: {bear_thesis.market_probability:.1%}
Edge: {bear_thesis.edge:.1%}
Core Argument: {bear_thesis.core_argument}
Catalysts: {', '.join(bear_thesis.catalysts)}
Failure Conditions: {', '.join(bear_thesis.failure_conditions)}
Supporting Signals: {', '.join(bear_thesis.supporting_signals)}

MARKET CONTEXT:
Question: {mbd.question}
Event Type: {mbd.event_type}
Current Probability: {mbd.current_probability:.1%}
Liquidity Score: {mbd.liquidity_score}/10
Volatility Regime: {mbd.volatility_regime}
Time to Resolution: {(mbd.expiry_timestamp - int(time.time())) / 86400:.1f} days

Execute all 5 cross-examination tests (evidence, causality, timing, liquidity, tail-risk) for BOTH theses.
For each test, provide: claim, challenge, outcome (survived/weakened/refuted), and score (-1.0/0.0/+1.0).
Then calculate overall scores for each thesis and identify key disagreements.
"""
        
        # Invoke LLM with structured output
        messages = [
            {"role": "system", "content": CROSS_EXAMINATION_PROMPT},
            {"role": "user", "content": prompt_context}
        ]
        
        logger.info("Invoking LLM for cross-examination analysis")
        result = await llm.ainvoke(messages)
        
        # Convert to DebateTest objects
        bull_tests = [
            DebateTest(
                test_type=test.test_type,
                claim=test.claim,
                challenge=test.challenge,
                outcome=test.outcome,
                score=test.score
            )
            for test in result.bull_tests
        ]
        
        bear_tests = [
            DebateTest(
                test_type=test.test_type,
                claim=test.claim,
                challenge=test.challenge,
                outcome=test.outcome,
                score=test.score
            )
            for test in result.bear_tests
        ]
        
        # Combine all tests
        all_tests = bull_tests + bear_tests
        
        # Use LLM-provided scores or calculate if needed
        bull_score = result.bull_score
        bear_score = result.bear_score
        
        # Use LLM-provided disagreements or generate if needed
        key_disagreements = result.key_disagreements
        if not key_disagreements:
            key_disagreements = identify_key_disagreements(
                bull_thesis, bear_thesis, bull_tests, bear_tests
            )
        
        # Create debate record
        debate_record = DebateRecord(
            tests=all_tests,
            bull_score=bull_score,
            bear_score=bear_score,
            key_disagreements=key_disagreements
        )
        
        duration_ms = int((time.time() - start_time) * 1000)
        
        logger.info(
            f"Cross-examination completed in {duration_ms}ms: "
            f"bull_score={bull_score:.2f}, bear_score={bear_score:.2f}, "
            f"disagreements={len(key_disagreements)}"
        )
        
        return {
            "debate_record": debate_record,
            "audit_log": [AuditEntry(
                stage="cross_examination",
                timestamp=int(time.time()),
                status="completed",
                details={
                    "duration_ms": duration_ms,
                    "bull_score": bull_score,
                    "bear_score": bear_score,
                    "bull_tests_passed": sum(1 for t in bull_tests if t.outcome == "survived"),
                    "bear_tests_passed": sum(1 for t in bear_tests if t.outcome == "survived"),
                    "total_tests": len(all_tests),
                    "disagreement_count": len(key_disagreements),
                    "score_difference": abs(bull_score - bear_score)
                }
            )]
        }
    
    except Exception as e:
        duration_ms = int((time.time() - start_time) * 1000)
        logger.error(f"Cross-examination failed after {duration_ms}ms: {e}")
        
        # Create fallback debate record with neutral scores
        fallback_tests = []
        for test_type in test_types:
            # Create neutral test for bull thesis
            fallback_tests.append(DebateTest(
                test_type=test_type,
                claim=f"Bull thesis claim for {test_type}",
                challenge=f"Unable to execute {test_type} test due to error",
                outcome="weakened",
                score=0.0
            ))
            # Create neutral test for bear thesis
            fallback_tests.append(DebateTest(
                test_type=test_type,
                claim=f"Bear thesis claim for {test_type}",
                challenge=f"Unable to execute {test_type} test due to error",
                outcome="weakened",
                score=0.0
            ))
        
        fallback_record = DebateRecord(
            tests=fallback_tests,
            bull_score=0.0,
            bear_score=0.0,
            key_disagreements=[f"Cross-examination error: {str(e)}"]
        )
        
        return {
            "debate_record": fallback_record,
            "audit_log": [AuditEntry(
                stage="cross_examination",
                timestamp=int(time.time()),
                status="failed",
                details={
                    "duration_ms": duration_ms,
                    "error": str(e),
                    "fallback_used": True
                }
            )]
        }


def create_cross_examination_node(config: EngineConfig):
    """
    Factory function to create cross-examination node with dependencies.
    
    Args:
        config: Engine configuration
        
    Returns:
        Async function that takes state and returns state update
        
    Examples:
        >>> config = load_config()
        >>> node = create_cross_examination_node(config)
        >>> result = await node(state)
    """
    async def node(state: GraphState) -> Dict[str, Any]:
        return await cross_examination_node(state, config)
    
    return node
